lib/wibox/widget/template.lua

```diff
@@ -0,0 +1,431 @@
+-- This is supposed to be a setter for an `update_now` property. This property
+-- however doesn't exist. We use this setter in the scope of the widget
+-- construction from wibox.widget internals to offer an easy way for the user to
+-- ask for the widget to be update right after its construction.
+-- Note : The update is not instantly called, but is registered as a normal
+-- update from the `:update()` method.
+-- @tparam[opt] boolean update_now Update the widget now.
+-- @method set_update_now
+-- @hidden
+function template:set_update_now(update_now)
+    if update_now then
+        self:update()
+    end
+end
+
+
+--- Create a new `wibox.widget.template` instance using the same template.
+--
+-- This copy will be blank. Note that `set_property`, `bind_property` or
+-- `update_callback` from `self` will **NOT** be transferred over to the copy.
+--
+-- The following example is a new widget to list a bunch of colors. It uses
+-- `wibox.widget.template` to let the module user define their own color
+-- widget. It does so by cloning the original template into new instances. This
+-- example doesn't handle removing or updating them to keep the size small.
+--
+--@DOC_wibox_widget_template_clone1_EXAMPLE@
+--
+-- @method clone
+-- @treturn wibox.widget.template The copy.
+function template:clone()
+    return template {
+        template = self._private.widget_template
+    }
+end
+
+--- Create a new `wibox.widget.template` instance.
+-- @tparam[opt] table args
+-- @tparam[opt] table|widget args.template The widget template to use.
+-- @tparam[opt] function args.update_callback The callback function to update
+--   the widget.
+-- @tparam[opt] boolean args.update_now Update the widget after its
+--   construction. This will call the `:update()` method with no parameter.
+-- @treturn wibox.widget.template The new instance.
+-- @constructorfct wibox.widget.template
+function template.new(args)
+    args = args or {}
+
+    local ret = wbase.make_widget(nil, nil, { enable_properties = true })
+
+    gtable.crush(ret, template, true)
+    ret._private.connections = {}
+
+    ret:set_template(args.template)
+    ret:set_update_callback(args.update_callback)
+    ret:set_update_now(args.update_now)
+
+    -- Apply the received buttons, visible, forced_width and so on
+    gtable.crush(ret, args)
+
+    rawset(ret, "_is_template", true)
+
+    return ret
+end
+
+--- Create a `wibox.widget.template` from a table.
+--
+-- @staticfct wibox.widget.template.make_from_value
+-- @tparam[opt=nil] table|wibox.widget.template|nil value A template declaration.
+-- @treturn wibox.widget.template The template object.
+function template.make_from_value(value)
+    if not value then return nil end
+
+    assert(
+        not rawget(value, "is_widget"),
+        "This property requires a widget template, not a widget object.\n"..
+        "Use `wibox.template` instead of `wibox.widget`"
+    )
+
+    if rawget(value, "_is_template") then return value end
```

Comment:
It means that with `local new_template = template.make_from_value(another_template)` `new_template` will be the same reference as `another_template`.

Maybe we should return `value:clone()` instead?